docs/design.md

```diff
@@ -0,0 +1,156 @@
+# Hypper Design
+
+Hypper is a package manager built on Helm and inspired by other package managers
+like zypper.
+
+## How Hypper Differs From Helm
+
+Why build a package manager on a package manager? Helm is a package manager that
+tries to make few assumptions. For example, Custom Resource Definitions (CRD)s
+are a Kubernetes resource that extend the Kubernetes API. They can be used in
+various ways and there are pitfalls with them. Helm makes very few assumptions
+when it comes to them. Hypper is going to make some different assumptions which
+will result in different features.
+
+As differences between Hypper and Helm arise they will be documented along with
+the reasons for the difference. This is in an effort to provide clarity.
+
+When features from Hypper can be upstreamed into Helm they will be. In that way,
+Hypper can act as a testing ground for Helm. We do not expect that every feature
+will fit in upstream Helm.
+
+A second element in the difference is that Helm, as a package manager, does not
+handle environment deployment operations. Those are left to higher level tools
+like Helmfile (a push model like Ansible) and the Flux Helm Operator (a pull
+model like Chef). Helm expects higher level tools will use Helm and charts for
+operating in different environments.
+
+Hypper will move will make some assumptions that move it more towards the
+environment management. This is in a different way from Helmfile or the Flux
+Helm Operator but further than Helm will go.
+
+## Directory Structure
+
+The directory structure is inspired by Helm with the following layout.
+
+```
+.
+├── cmd
+│   └── hypper
+├── docs
+└── pkg
+    └── action 
+    └── ... 
+```
+
+- The _cmd/hypper_ directory contains the client application. This CLI for mac,
+  linux, and windows is both a functional application and reference application
+  of the library.
+- _docs_ contains documentation on the design and usage of hypper.
+- The _pkg_ directory contains a library meant to be imported by _cmd/hypper_ and
+  other applications. The functionality needed is implemented here.
+- _pkg/actions_ contains the main functionality used by _cmd/hypper_.
+
+This is the same layout as Helm itself.
+
+## Design Intent
+
+The following outline the initial major intentions of hypper.
+
+### Modern Terminal UI
+
+Colors and icons are often present in modern terminals. Sometimes referred to
+as eye candy, these features can improve user experience by drawing attention to
+important things.
+
+### Optional Dependencies
+
+In Helm there are two ways to handle dependencies.
+
+1. Explicitly declare other charts as dependencies. This will cause those charts
+   to be installed. When the templates are rendered they happen in the same
+   sandboxed environment.
+2. Install charts separately and use values to connect them to each other. For
```

Comment:
How do we plan to store this information across runs? Are we gonna use a CRD for hypper? local database? on-the-fly search?

How are we expecting this to work? Automatically? Will it require to put restrictions in place in case there is a existing X that we **dont** want other charts to use?

Reply:
This section of the document is talking about Helm rather than hypper. The design of this was all through through so I can talk to the details.

> How do we plan to store this information across runs? Are we gonna use a CRD for hypper? local database? on-the-fly search?

CRDs are an extension of the Kubernetes API. Extending the API is something that's usually only reserved for local dev and cluster admins. Most app operators don't have this kind of access and shouldn't. So, Helm does not assume this access.  We talked to people at companies about it.

Helm uses Secrets as the default with a configurable option to use PostgreSQL. A database is hard because you have to have that setup first. Where does it live? How is access handled? There is a fair amount of setup before you can get going. It's not like using an easy button.

With hypper we are working with the Helm library and Helm charts. I'm hoping with can piggy back on the existing storage setup Helm uses. That way we don't have to invent something new. I think we would look at a change when we encountered a need for it and then we'd have to think through the holistic experience.

If you're not familiar with how Helm does things I would suggest looking at it. I'm happy to answer any questions.

> How are we expecting this to work? Automatically? Will it require to put restrictions in place in case there is a existing X that we dont want other charts to use?

We can store metadata in Chart annotations. For example, have a default name and namespace to look for it in. All of the charts in a repo can be setup to look for each other. If the user installing has access to the namespace and resource we can connect them up. This would work with RBAC.

If you look at Rancher charts right now you'll see the use of annotations to store data. [For example, see this one](https://github.com/rancher/charts/blob/7a05185008739bd29da34f68b1d19027ba30287f/charts/rancher-logging/Chart.yaml).

Reply:
:+1: